src/HydraAuction/Tx/FeeEscrow.hs

```diff
@@ -0,0 +1,118 @@
+
+-- Hydra imports
+import Hydra.Cardano.Api (
+  Lovelace (..),
+  Value,
+  lovelaceToValue,
+  txOutValue,
+  valueToLovelace,
+  pattern ReferenceScriptNone,
+  pattern TxMintValueNone,
+  pattern TxOut,
+  pattern TxOutDatumNone,
+ )
+
+-- Hydra auction imports
+import HydraAuction.OnChain (AuctionScript (..))
+import HydraAuction.Tx.Common (
+  scriptPlutusScript,
+  scriptUtxos,
+ )
+import HydraAuction.Types (
+  AuctionTerms (..),
+  FeeEscrowRedeemer (..),
+ )
+import HydraAuctionUtils.L1.Runner (L1Runner)
+import HydraAuctionUtils.Monads (
+  fromPlutusAddressInMonad,
+  logMsg,
+ )
+import HydraAuctionUtils.Monads.Actors (
+  actorTipUtxo,
+  addressAndKeys,
+ )
+import HydraAuctionUtils.Tx.AutoCreateTx (
+  AutoCreateParams (..),
+  autoSubmitAndAwaitTx,
+ )
+import HydraAuctionUtils.Tx.Build (
+  mkInlinedDatumScriptWitness,
+ )
+import HydraAuctionUtils.Tx.Utxo (
+  filterAdaOnlyUtxo,
+ )
+
+distributeFee :: AuctionTerms -> L1Runner ()
+distributeFee terms = do
+  logMsg "Distributing fees to delegates"
+
+  (actorAddress, _, actorSk) <- addressAndKeys
+
+  actorMoneyUtxo <- filterAdaOnlyUtxo <$> actorTipUtxo
+
+  feeEscrowUtxo <- scriptUtxos FeeEscrow terms
+
+  feeEscrowValue <- case UTxO.pairs feeEscrowUtxo of
+    [(_, txOut)] -> pure $ txOutValue txOut
+    _ -> fail "wrong number of utxos in fee escrow script"
+
+  lovelaceAmt <- case valueToLovelace feeEscrowValue of
+    Just (Lovelace l) -> pure l
+    Nothing -> fail "fee escrow asset does not contain ada-only"
+
+  delegateOuts <- mapM mkFeeOut (zipDistributingValue lovelaceAmt (delegates terms))
+
+  logMsg $ show delegateOuts
+  void $
+    autoSubmitAndAwaitTx $
+      AutoCreateParams
+        { signedUtxos = [(actorSk, actorMoneyUtxo)]
+        , additionalSigners = []
+        , referenceUtxo = mempty
+        , witnessedUtxos = [(feeWitness, feeEscrowUtxo)]
+        , collateral = Nothing
+        , outs = delegateOuts
+        , toMint = TxMintValueNone
+        , changeAddress = actorAddress
+        , validityBound = (Nothing, Nothing)
+        }
+  where
+    feeScript = scriptPlutusScript FeeEscrow terms
+    feeWitness = mkInlinedDatumScriptWitness feeScript DistributeFees
+
+    mkFeeOut (delegatePKH, valuePerDelegate) = do
+      delegateAddress <- fromPlutusAddressInMonad $ pubKeyHashAddress delegatePKH
+      pure $
+        TxOut
+          delegateAddress
+          valuePerDelegate
+          TxOutDatumNone
+          ReferenceScriptNone
+
+zipDistributingValue :: Integer -> [PubKeyHash] -> [(PubKeyHash, Value)]
+zipDistributingValue lovelaceAmt delegatePKHs = zip delegatePKHs (lovelaceToValue . Lovelace <$> distributedLovelace)
+  where
+    delegateNumber = length delegatePKHs
+    (quotient, remainder) = divMod lovelaceAmt (toInteger delegateNumber)
+
+    spread _ [] = []
+    spread 0 xs = xs
+    spread k (x : xs) = (x + 1) : spread (k - 1) xs
+
+    distributedLovelace = spread remainder (replicate delegateNumber quotient)
```

Comment:
This is not needed because of `auctionFeePerDelegate` too

Reply:
Or is all this code due to #298 ? In that case please put comments for `spread` function and write about this motivation in spec.

Reply:
If we leave this kind of logic, it should have property-based unit test matching it with on-chain check.

So I think it may be simpler to just reuse this function in on-chain too. That way there is no need to test that they match each over. And it seems like there is no use for indeterminacy of remainder spreading anyway.

While this needs unit-test or property-based test anyway.

Reply:
> Or is all this code due to https://github.com/mlabs-haskell/hydra-auction/issues/298 ? I

No, this is only distributing ADA to the delegates in a balanced way.

> If we leave this kind of logic, it should have property-based unit test matching it with on-chain check.

> So I think it may be simpler to just reuse this function in on-chain too. That way there is no need to test that they match each over. And it seems like there is no use for indeterminacy of remainder spreading anyway.

> While this needs unit-test or property-based test anyway.

I don't think we really need unit tests for this helper function, its quite a trivial implementation, we already have on and offchain tests for this anyway, so we are testing it.

Also, I don't think it would be nice to use this on-chain, this gives slightly more ADA to the delegates appearing first in the list. In the on-chain we should only verify that each delegates receives >= `auctionFeePerDelegate`.
The off-chain has to actually consume the utxo, so in case there is more ada than `auctionFeePerDelegate * length delegates` it can decide how to distribute the extra ada

Reply:
> No, this is only distributing ADA to the delegates in a balanced way.

I mean that if we do are not concerned with #298 then this ADA separation code is not needed and just `auctionFeePerDelegate` can be used.

> I don't think we really need unit tests for this helper function, its quite a trivial implementation

Integer division and such logic is exactly the kind of stuff where people get errors like off-by-one or something like that. Also we already got off-by-one errors in intervals helpers code.

> we already have on and offchain tests for this anyway, so we are testing it.

Our tests do not test ADA amount at all and this is an accident waiting to be happen.

> this gives slightly more ADA to the delegates appearing first in the list

Yes, and your specification says that this is okay. So why accept different remainder distributions then if spec says that any of them are good?

> The off-chain has to actually consume the utxo, so in case there is more ada than auctionFeePerDelegate * length delegates it can decide how to distribute the extra ada

What is `off-chain`? Why where can be extra ADA? If it is winning lot, then this does not interfere with delegate ADA distribution, cuz bidder cannot be an delegate.

Reply:
@GeorgeFlerovsky Can delegate be a bidder too?